node/narwhal/src/gateway.rs

```diff
@@ -877,30 +888,34 @@ impl<N: Network> Handshake for Gateway<N> {
             self.handshake_inner_responder(peer_addr, &mut peer_ip, stream).await
         }
         .map_err(|error| {
-            // If the handshake failed, remove the connecting peer.
+            // If the handshake failed, remove the connecting peer and the noise state.
             if let Some(ip) = peer_ip {
                 self.connecting_peers.lock().shift_remove(&ip);
+                self.noise_states.write().remove(&ip);
             }
 
             error
         })?;
 
         // Check for double connects.
         match self.noise_states.write().entry(*peer_ip) {
-            Occupied(_) => {
-                // If the entry already exists, the handshake was already performed. This can
-                // happen if there is a poorly timed double-connect attempt. In this case, we
-                // simply drop the connection and cancel the handshake.
-                self.connecting_peers.lock().shift_remove(peer_ip);
-                warn!("{CONTEXT} Handshake with '{peer_ip}' failed: noise state already exists");
-                return Err(io::ErrorKind::AlreadyExists.into());
-            }
-            Vacant(entry) => {
-                // Otherwise, we store the noise state for reuse by the Reading and Writing impls
-                // and proceed.
+            Occupied(mut entry) => {
+                // A connection with this peer listener has already occured, cancel the handshake.
+                if entry.get().is_some() {
+                    self.connecting_peers.lock().shift_remove(peer_ip);
```

Comment:
this should be unnecessary, as it's done in `perform_handshake` in case of an early return here

Reply:
nvm, I mixed the sites :ok_hand: